### PR TITLE
Derive default host/advertised host without Dockerfile environment variables

### LINKED ISSUE #19114: startup.sh script overrides helm config
**Description**

I am trying to set the following in the helm [configmap](https://github.com/camunda/camunda-platform-helm/pull/1910) for the zeebeGateway:
```
 zeebe:
      gateway:
          host: 0.0.0.0
```

The [startup.sh](https://github.com/camunda/camunda/blob/main/zeebe/docker/utils/startup.sh) script overrides this config with the use of env vars:
```
HOST=$(hostname -i)

if [ "$ZEEBE_STANDALONE_GATEWAY" = "true" ]; then
    export ZEEBE_GATEWAY_NETWORK_HOST=${ZEEBE_GATEWAY_NETWORK_HOST:-${HOST}}
    export ZEEBE_GATEWAY_CLUSTER_HOST=${ZEEBE_GATEWAY_CLUSTER_HOST:-${ZEEBE_GATEWAY_NETWORK_HOST}}

    exec /usr/local/zeebe/bin/gateway

```

We could remove these env vars in the startup.sh script and let the helm configmap have control. I am aware that this might cause potential problems running zeebe outside of helm.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR removes the explicit configuration environment variables set in `startup.sh` and in the `Dockerfile`. Note that these still have some explicit environment variables, but they are not related to the Spring application configuration, and instead are used only in the script itself to pick which application to start (to be removed eventually, but out of scope here).

The motivation here is that when we set these explicit environment variables, the only way for a user to reconfigure them is by setting the same environment variable - changing the configuration file or setting a system property will not work due to configuration source precedence rules.

As such, it can lead to confusion for users because they will not know that the startup.sh script sets these things.

> [!Note]
> The motivation behind this was that, historically, you wanted to be able to launch a Docker cluster with minimal configuration. Since we split the host/advertised host in order to support reverse proxy set ups (e.g. using service meshes), where the bind host may be different from the host on which other nodes should reply to, we wanted a minimal way to launch a container without having to always specify everything. If we did not specify a host, then nodes would tell other nodes to reply on 0.0.0.0, which would make nodes reply, well, to themselves. So we opted to always set the node's IP address as the bind host by default in Docker.

To make this work in a backwards compatible way, we implemented a few changes:

1. Remove configuration related environment variables from the `Dockerfile` and the `startup.sh` script.
1. We don't explicitly set the `0.0.0.0` as default bind host in the `application.yaml` file. Since it was explicitly set, we would then use it as the default advertised host (without the environment variables), which would mean you _always_ have to specify the advertised host if you want to do clustering.
1. We don't set default `host` and `advertisedHost` property values in the `NetworkCfg` (broker) and `ClusterCfg` (gateway) classes. This lets us distinguish if the user has explicitly set these properties or not. Instead, we return appropriate default based on simple heuristics:
  - For `host`:
    - If `host` is explicitly set, use that.
    - If not, use the appropriate wildcard address (`0.0.0.0` or `::`) based on preferred IP stack.
  - For `advertisedHost`:
    - If `advertisedHost` is explicitly set, use that.
    - If `host` is explicitly set, use that.
    - If not, then use a derived default advertised host.

The derived default advertised host (returned by `Address#defaultAdvertisedHost`) is computed in the following way:
- Return `InetAddress#getLocalHost`: this essentially grabs the host name of the node (so equivalent to the `hostname` command on Linux), and resolves its IP address.
- If that fails, go through _all_ network interfaces and grab the first non-loopback address, respecting the IP stack preference.
- If there is none, return the appropriate loopback address for the IP stack preference.

This is almost 100% backwards compatible with the previous solution, meaning:
- Since `hostname -i` was returning the IP address of the hostname (in fact, it returns _all_ IP addresses, which was sometimes a problem), the default should still hold as we will also resolve that.
- The Helm charts explicitly set both host and advertised host, so no changes required there.
- Users already explicitly setting the host to 0.0.0.0 and not setting any advertised host were already advertising 0.0.0.0 anyway, so this doesn't break that.

It is a _breaking_ change in the following edge case:
- Users who were using the default `0.0.0.0` host for both binding and advertising, and for whom the derived host will behave differently.

I think that's an extremely narrow edge case, and it's acceptable here.

## Related issues

closes #19114

### CODE REVIEW

**Comment by npepinpe** [comment]
@Zelldon - only adding you so you're aware of this, you don't need to review

**Review by lenaschoenburg — APPROVED** [review]
Very nice to review :+1:

File: dist/src/main/config/gateway.yaml.template [hunk 1]
```diff
@@ -87,10 +87,15 @@
       # This setting can also be overridden using the environment variable ZEEBE_GATEWAY_CLUSTER_PORT.
       # port: 26502
 
-      # Controls the advertised host; if omitted defaults to the host. This is particularly useful if your
-      # gateway stands behind a proxy.
+      # Controls the advertised host. This is particularly useful if your broker stands behind a
+      # proxy. If not set, its default is computed as:
+      #   - If zeebe.gateway.network.host was explicitly and it is not the ANY address (e.g. "0.0.0.0" or "::"), use this.
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply dist/src/main/config/gateway.yaml.template:92]: oops :smile:

File: zeebe/atomix/utils/src/main/java/io/atomix/utils/net/AddressInitializations.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      return NetUtil.LOCALHOST6;
+    }
+
+    return NetUtil.LOCALHOST4;
+  }
+
+  private static NonLoopBackAddresses collectNonLoopBackAddresses(
+      final Stream<InetAddress> networkInterfaces, final boolean isIPv6Preferred) {
+    // loop through all network interfaces' addresses, and pick the first non loop back IPv4 and
+    // IPv6 addresses; there is no combiner as this is not meant to be used with parallel streams
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply zeebe/atomix/utils/src/main/java/io/atomix/utils/net/AddressInitializations.java:59]: :exclamation: Oh, I knew about `parallel()`, but makes sense that there is also a `sequential()`

File: dist/src/main/config/broker.standalone.yaml.template [hunk 3]
```diff
@@ -198,10 +204,13 @@
         # This setting can also be overridden using the environment variable ZEEBE_BROKER_NETWORK_COMMANDAPI_PORT.
         # port: 26501
 
-        # Controls the advertised host; if omitted defaults to the host. This is particularly useful if your
-        # broker stands behind a proxy.
+        # Controls the advertised host. This is particularly useful if your broker stands behind a proxy.
+        # If omitted, defaults to:
+        #   - If zeebe.broker.network.commandApi.host was set, use this.
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply dist/src/main/config/broker.standalone.yaml.template:209]: At the moment, yes. Like with the general `zeebe.broker.network.host`, if you _explicitly_ set it to 0.0.0.0 (so it's not using the default value), the advertised host will _also_ be 0.0.0.0. The reasoning is that in most cases, when you explicitly set a host, this is also what you want to advertise, and _most_ cases where you set 0.0.0.0 but want to advertise something else, you would also explicitly set the advertised host.

[the remark this answers is not in the document]
Comment (lenaschoenburg) [reply dist/src/main/config/broker.standalone.yaml.template:209]: Ah, are the docs for `ZEEBE_BROKER_NETWORK_ADVERTISEDHOST` wrong then?
> If zeebe.broker.network.host was explicitly and it is not the ANY address (e.g. "0.0.0.0" or "::"), use this.

[the remark this answers is not in the document]
Comment (npepinpe) [reply dist/src/main/config/broker.standalone.yaml.template:209]: Oops, correct, that was a leftover :sweat_smile:

**Pull request merged**